Duplicate protocol negotiation message

I've been testing protocol 3.2 and I'm experiencing an interesting bug with pgbouncer.

I've currently only tested using the bitnamilegacy docker image, I will try to build from source to try again but here's my repro

```sh
docker network create pg
docker run -d --name postgres -e POSTGRES_PASSWORD=password --network=pg postgres:18-bookworm
docker run -d --name pgbouncer -e POSTGRESQL_PASSWORD=password -e PGBOUNCER_AUTH_USER=postgres -e POSTGRESQL_HOST=postgres -p 6432:6432 --network=pg bitnamilegacy/pgbouncer:1.24.1-debian-12-r10

docker exec -it postgres psql -U postgres -c "CREATE ROLE example WITH LOGIN PASSWORD 'password';"

PGMAXPROTOCOLVERSION=3.2 psql "postgresql://example:password@localhost:6432/postgres"
```

Here's the error message I get back

```
psql: error: connection to server at "localhost" (::1), port 6432 failed: Connection refused
        Is the server running on that host and accepting TCP/IP connections?
connection to server at "localhost" (127.0.0.1), port 6432 failed: received duplicate protocol negotiation message
```

I can see in wireshark that pgbouncer is indeed returning two identical protocol negotiation messages.

Interestingly enough, connecting to `"postgresql://postgres:password@localhost:6432/postgres"` does not trigger the same error - only one protocol negotiation message is sent.

---

I have also tested a 1.25.1 image curtesy of [our bitnami image fork](https://hub.docker.com/layers/neondatabase/pgbouncer/1.25.1-c8ae4f5a273e-21147800173/images/sha256-ea9745213cee29e585302d238777c050c2cc3d912222ae5eddf410f6fb0c7d65). I also tested 1.23.0, which is the version of pgbouncer that first implemented protocol negotiation, which also has the same behaviour - this is not a new regression.

Managed to write a pytest for this (had to change `psycopg` to `psycopg[binary]`).

```python
def test_protocol_negotiation(bouncer, pg):
    config = f"""
    [databases]
    postgres = host={bouncer.pg.host} port={bouncer.pg.port} 

    [pgbouncer]
    listen_port = {bouncer.port}
    listen_addr = {bouncer.host}
    auth_file = {bouncer.auth_path}
    auth_type = scram-sha-256
    auth_user = postgres
    logfile = {bouncer.log_path}
    """

    pg.sql("set password_encryption = 'scram-sha-256'; CREATE ROLE example WITH LOGIN PASSWORD 'password'")

    with bouncer.run_with_config(config):
        with bouncer.cur(user="example", password="password", dbname="postgres", max_protocol_version="3.2") as cur:
            cur.execute("SELECT 1")
```

```
FAILED test/test_misc.py::test_protocol_negotiation - psycopg.OperationalError: connection failed: connection to server at "127.0.0.1", port 10202 failed: received duplicate protocol negotiation message
```

--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -27,7 +27,7 @@ task:
       env:
         ENABLE_VALGRIND: yes
         CFLAGS: -O0 -g
-        PGVERSION: 17
+        PGVERSION: 18
     - container:
         image: ubuntu:22.04
       env:

--- a/include/bouncer.h
+++ b/include/bouncer.h
@@ -689,6 +689,7 @@ struct PgSocket {
 
 	bool wait_for_welcome : 1;	/* client: no server yet in pool, cannot send welcome msg */
 	bool welcome_sent : 1;		/* client: client has been sent the welcome msg */
+	bool protocol_negotiated : 1;	/* client: NegotiateProtocolVersion already sent */
 	bool wait_for_user_conn : 1;	/* client: waiting for auth_conn server connection */
 	bool wait_for_user : 1;		/* client: waiting for auth_conn query results */
 	bool wait_for_auth : 1;		/* client: waiting for external auth (PAM/LDAP) to be completed */

--- a/src/client.c
+++ b/src/client.c
@@ -1093,8 +1093,8 @@ static bool decide_startup_pool(PgSocket *client, PktHdr *pkt)
 			unsupported_protocol_extensions.write_pos
 			);
 		res = pktbuf_send_immediate(buf, client);
+		pktbuf_free(buf);
 		if (!res) {
-			pktbuf_free(buf);
 			disconnect_client(client, false, "unable to send protocol negotiation packet");
 			return false;
 		}

--- a/src/client.c
+++ b/src/client.c
@@ -1080,10 +1080,12 @@ static bool decide_startup_pool(PgSocket *client, PktHdr *pkt)
 		}
 	}
 
-	if (pkt->type == PKT_STARTUP_V3_UNSUPPORTED || unsupported_protocol_extensions_count > 0) {
+	if ((pkt->type == PKT_STARTUP_V3_UNSUPPORTED || unsupported_protocol_extensions_count > 0)
+	    && !client->protocol_negotiated) {
 		PktBuf *buf = pktbuf_dynamic(512);
 		int res;
 
+		client->protocol_negotiated = true;
 		pktbuf_write_NegotiateProtocolVersion(
 			buf,
 			unsupported_protocol_extensions_count,

--- a/test/test_auth.py
+++ b/test/test_auth.py
@@ -1206,6 +1206,41 @@ def test_auth_query_no_set_commands(bouncer, pg):
         pg.sql("DROP FUNCTION IF EXISTS auth_check_search_path(TEXT)")
 
 
+@pytest.mark.md5
+@pytest.mark.skipif(
+    "psycopg.pq.version() < 180000", reason="libpq 18+ required for protocol 3.2"
+)
+def test_auth_query_protocol_negotiation(bouncer, pg):
+    """
+    Test that protocol version negotiation works correctly with auth_query.
+    This is a regression test for GitHub issue #1459 where clients connecting
+    with protocol version 3.2 would receive duplicate NegotiateProtocolVersion
+    messages when auth_query was needed.
+
+    The bug occurred because decide_startup_pool() sends the
+    NegotiateProtocolVersion message, but when auth_query is needed, the
+    startup packet would be reprocessed after auth_query completes, potentially
+    calling decide_startup_pool() again and sending a duplicate message.
+    """
+    bouncer.default_db = "authdb"
+    bouncer.admin("set auth_type='md5'")
+
+    # First, kill all server connections to ensure a cold pool. This is
+    # important because the bug only manifests when auth_query needs to
+    # actually run (first connection), not when user is already cached.
+    bouncer.admin("reconnect authdb")
+
+    # If the bug is present, this will fail with
+    # "received duplicate protocol negotiation message"
+    bouncer.sql(
+        "SELECT 1",
+        dbname="authdb",
+        user="someuser",
+        password="anypasswd",
+        max_protocol_version="3.2",
+    )
+
+
 @pytest.mark.skipif("WINDOWS", reason="We do not expect to support ldap on Windows")
 @pytest.mark.skipif(not LDAP_SUPPORT, reason="pgbouncer is built without LDAP support")
 def test_ldap_auth(bouncer_with_openldap):

